Feat/parity chainspec, tests - kensington ready (via parity chainspec)

meowsbits opened the pull request:
This builds from and supersedes #97 . 

It creates a flag `--chainspec.parity=""` which can read a parity-style chainspec JSON file.

I just ran a successful fast sync on Kensington:

(attached is the corresponding JSON file, ala https://github.com/eth-classic/kensington/blob/master/parity.json, but [with a minor adjustment to make genesis account balance vals strings](https://github.com/eth-classic/kensington/pull/11
))

[kensington-parity.txt](https://github.com/multi-geth/multi-geth/files/3278735/kensington-parity.txt) (it's really json, but github only allows .txt files)


```
$ make geth && rm -rf ken && ./build/bin/geth --chainspec.parity=./params/kensington-parity.json --datadir ./ken                                        
build/env.sh go run build/ci.go install ./cmd/geth                                                                                                      
>>> /home/ia/go1.12.1.linux-amd64/go/bin/go install -ldflags -X main.gitCommit=b1d1fdcc2425a6f21394bbd646c7925b381a8591 -X main.gitDate=20190611 -v ./cmd/geth
github.com/ethereum/go-ethereum/cmd/utils                                                                                                                                  
github.com/ethereum/go-ethereum/cmd/geth                                                                                                                
Done building.                                                                                                                                          
Run "/home/ia/go/src/github.com/ethereum/go-ethereum/build/bin/geth" to launch geth.                                                                                    
INFO [06-11|16:48:16.157] Bumping default cache on mainnet         provided=1024 updated=4096                                                                                                                                                                                     
INFO [06-11|16:48:16.158] Maximum peer count                       ETH=50 LES=0 total=50                                                   
INFO [06-11|16:48:16.158] Smartcard socket not found, disabling    err="stat /run/pcscd/pcscd.comm: no such file or directory"                          
WARN [06-11|16:48:16.159] at leaset ist set                        chainspecpar=nil LOG15_ERROR="Normalized odd number of arguments by adding nil"
INFO [06-11|16:48:16.159] Reading chainspec (parity format)        ./params/kensington-parity.json=nil LOG15_ERROR="Normalized odd number of arguments by adding nil"
INFO [06-11|16:48:16.159] Starting peer-to-peer node               instance=MultiGeth/v1.9.0-unstable-b1d1fdcc-20190611/linux-amd64/go1.12.1
INFO [06-11|16:48:16.159] Allocated trie memory caches             clean=1024.00MiB dirty=1024.00MiB                                                    
INFO [06-11|16:48:16.159] Allocated cache and file handles         database=/home/ia/go/src/github.com/ethereum/go-ethereum/ken/geth/chaindata cache=2.00GiB handles=524288
INFO [06-11|16:48:16.195] Opened ancient database                  database=/home/ia/go/src/github.com/ethereum/go-ethereum/ken/geth/chaindata/ancient   
INFO [06-11|16:48:16.195] Writing custom genesis block                                                                                     
INFO [06-11|16:48:16.196] Persisted trie from memory database      nodes=12 size=1.74KiB time=61.929µs gcnodes=0 gcsize=0.00B gctime=0s livenodes=1 livesize=0.00B
INFO [06-11|16:48:16.196] Initialised chain configuration          config="{ChainID: 1337 Homestead: 0 DAO: <nil> DAOSupport: false EIP150: 0 EIP155: 0 EIP158: <nil> Byzantium: <nil> Disposal: 0 ECIP1017: 2000000 EIP160: 0 ECIP1010PauseBlock: 0 ECIP1010Length: 2000000 Const
antinople: <nil> ConstantinopleFix: <nil> Engine: ethash}"                                                                                                                                                                      
INFO [06-11|16:48:16.196] Disk storage enabled for ethash caches   dir=/home/ia/go/src/github.com/ethereum/go-ethereum/ken/geth/ethash count=3       
INFO [06-11|16:48:16.196] Disk storage enabled for ethash DAGs     dir=/home/ia/.ethash count=2                                                          
INFO [06-11|16:48:16.196] Initialising Ethereum protocol           versions="[63 62]" network=88 dbversion=<nil>                                        
WARN [06-11|16:48:16.196] Upgrade blockchain database version      from=<nil> to=7                                                                       
INFO [06-11|16:48:16.222] Loaded most recent local header          number=0 hash=5ab4c8…5d9420 td=131072 age=50y1mo4w                                    
INFO [06-11|16:48:16.222] Loaded most recent local full block      number=0 hash=5ab4c8…5d9420 td=131072 age=50y1mo4w                                  
INFO [06-11|16:48:16.222] Loaded most recent local fast block      number=0 hash=5ab4c8…5d9420 td=131072 age=50y1mo4w                                   
INFO [06-11|16:48:16.222] Regenerated local transaction journal    transactions=0 accounts=0                                                            
INFO [06-11|16:48:16.239] Allocated fast sync bloom                size=2.00GiB                                                                         
INFO [06-11|16:48:16.241] Initialized fast sync bloom              items=12 errorrate=0.000 elapsed=803.889µs                                                                            
INFO [06-11|16:48:16.252] New local node record                    seq=1 id=7580988b61d4d639 ip=127.0.0.1 udp=30303 tcp=30303                           
INFO [06-11|16:48:16.252] Started P2P networking                   self=enode://29f12063130ed25359b79ceb04d2596e26aa8842db1cdaf5a666e04931332f0e8c36d7b18270dfb56dc0e0bc82c3ddd2e9f1eddae6763b605ebc80846986a4a7@127.0.0.1:30303
INFO [06-11|16:48:16.253] IPC endpoint opened                      url=/home/ia/go/src/github.com/ethereum/go-ethereum/ken/geth.ipc                     
INFO [06-11|16:48:16.260] Mapped network port                      proto=tcp extport=30303 intport=30303 interface=NAT-PMP(10.0.1.1)                                                     
INFO [06-11|16:48:16.265] Mapped network port                      proto=udp extport=30303 intport=30303 interface=NAT-PMP(10.0.1.1)                          
INFO [06-11|16:48:16.361] New local node record                    seq=2 id=7580988b61d4d639 ip=71.226.173.63 udp=30303 tcp=30303                       
INFO [06-11|16:48:46.252] Block synchronisation started                                                                                                 
INFO [06-11|16:48:48.631] Imported new state entries               count=5 elapsed=73.616µs  processed=5 pending=0 retry=0 duplicate=0 unexpected=0     
INFO [06-11|16:48:49.040] Imported new block headers               count=192 elapsed=508.289ms number=192 hash=b0068d…02aaa1 age=1d8h10m                                 
INFO [06-11|16:48:49.045] Imported new block receipts              count=192 elapsed=778.368µs number=192 hash=b0068d…02aaa1 age=1d8h10m  size=768.00B  
INFO [06-11|16:48:49.053] Imported new block headers               count=192 elapsed=11.495ms  number=384 hash=50345c…c9ec97 age=1d8h2m                  
...
INFO [06-11|16:49:00.482] Imported new block headers               count=192 elapsed=12.722ms  number=9984 hash=4597ab…67995c age=8h28m33s
INFO [06-11|16:49:00.611] Imported new block receipts              count=165 elapsed=3.512ms   number=9903 hash=9872f6…c1b0ef age=8h55m57s  size=1.14KiB
INFO [06-11|16:49:00.798] Imported new block headers               count=192 elapsed=13.854ms  number=10176 hash=03a727…4997e5 age=7h31m52s
INFO [06-11|16:49:00.918] Imported new block receipts              count=273 elapsed=5.244ms   number=10176 hash=03a727…4997e5 age=7h31m52s  size=1.57KiB
INFO [06-11|16:49:01.033] Imported new block headers               count=192 elapsed=13.767ms  number=10368 hash=2e13ea…b5e16d age=6h34m29s
INFO [06-11|16:49:01.036] Imported new block receipts              count=4   elapsed=59.932µs  number=10180 hash=3231ec…89eb20 age=7h30m59s  size=16.00B
INFO [06-11|16:49:01.375] Imported new block headers               count=192 elapsed=40.922ms  number=10560 hash=1487fc…d00413 age=5h34m40s
INFO [06-11|16:49:01.440] Imported new block receipts              count=276 elapsed=1.213ms   number=10456 hash=cff9d5…ce1865 age=6h2m33s   size=1.58KiB
INFO [06-11|16:49:01.564] Imported new block headers               count=192 elapsed=13.137ms  number=10752 hash=1885d6…5019d9 age=4h38m20s
INFO [06-11|16:49:01.667] Imported new block receipts              count=222 elapsed=5.107ms   number=10678 hash=cecf61…2ffcd6 age=4h57m45s  size=1.87KiB
INFO [06-11|16:49:01.861] Imported new block headers               count=192 elapsed=93.020ms  number=10944 hash=b433c2…ec116e age=3h46m11s
INFO [06-11|16:49:01.871] Imported new block receipts              count=81  elapsed=461.197µs number=10759 hash=62df47…c6e1a9 age=4h36m37s  size=1.32KiB
INFO [06-11|16:49:01.990] Imported new block receipts              count=168 elapsed=3.630ms   number=10927 hash=580576…89b2f4 age=3h51m31s  size=1.66KiB
INFO [06-11|16:49:02.206] Imported new state entries               count=2   elapsed=145.857µs processed=7 pending=0 retry=0 duplicate=0 unexpected=0
INFO [06-11|16:49:02.215] Imported new block receipts              count=1   elapsed=161.104µs number=10928 hash=090d26…67c797 age=3h50m37s  size=4.00B
INFO [06-11|16:49:02.215] Committed new head block                 number=10928 hash=090d26…67c797
INFO [06-11|16:49:02.255] Deallocated fast sync bloom              items=19 errorrate=0.000
INFO [06-11|16:49:02.272] Imported new chain segment               blocks=16 txs=0 mgas=0.000 elapsed=15.225ms  mgasps=0.000 number=10944 hash=b433c2…ec116e age=3h46m12s  dirty=10.64KiB
INFO [06-11|16:49:02.272] Upgrading chain index                    type=bloombits percentage=0
INFO [06-11|16:49:02.517] Finished upgrading chain index           type=bloombits
INFO [06-11|16:49:05.018] Imported new block headers               count=48  elapsed=27.664ms  number=10992 hash=fd39ee…2bb6e3 age=2h34m57s
INFO [06-11|16:49:05.024] Imported new chain segment               blocks=48 txs=0 mgas=0.000 elapsed=5.384ms   mgasps=0.000 number=10992 hash=fd39ee…2bb6e3 age=2h34m57s  dirty=42.56KiB
INFO [06-11|16:49:05.096] Fast sync complete, auto disabling
^CINFO [06-11|16:49:37.984] Got interrupt, shutting down...
INFO [06-11|16:49:37.985] IPC endpoint closed                      url=/home/ia/go/src/github.com/ethereum/go-ethereum/ken/geth.ipc
INFO [06-11|16:49:37.985] Writing cached state to disk             block=10992 hash=fd39ee…2bb6e3 root=5473b3…a0e1bb
INFO [06-11|16:49:37.985] Persisted trie from memory database      nodes=2  size=489.00B  time=254.011µs gcnodes=0 gcsize=0.00B gctime=0s livenodes=127 livesize=30.08KiB
INFO [06-11|16:49:37.985] Writing cached state to disk             block=10991 hash=05f1ea…95b55d root=c00204…41df1d
INFO [06-11|16:49:37.986] Persisted trie from memory database      nodes=2  size=489.00B  time=147.283µs gcnodes=0 gcsize=0.00B gctime=0s livenodes=125 livesize=29.61KiB
INFO [06-11|16:49:37.986] Writing cached state to disk             block=10865 hash=dc6e2b…df18f8 root=5b665a…c15244
INFO [06-11|16:49:37.986] Persisted trie from memory database      nodes=0  size=0.00B    time=6.189µs   gcnodes=0 gcsize=0.00B gctime=0s livenodes=125 livesize=29.61KiB
INFO [06-11|16:49:37.987] Blockchain manager stopped
INFO [06-11|16:49:37.987] Stopping Ethereum protocol
INFO [06-11|16:49:37.988] Ethereum protocol stopped
INFO [06-11|16:49:37.988] Transaction pool stopped
```

meowsbits committed:
tests: copy parity chainspecs to tests dir

meowsbits committed:
params: add missing ropsten dao fork block

meowsbits committed:
consensus,params: introduce config maps for reward and dbomb delays

meowsbits committed:
Merge branch 'master' of https://github.com/multi-geth/multi-geth into feat/parity-chainspec-tests

meowsbits committed:
cmd/puppeth,consensus/ethash,params: implement block reward schedule

meowsbits committed:
consensus/ethash,core/rawdb,params: fix difficulty logic, add tests, marshaling methods

meowsbits committed:
consensus/ethash: fix comment and useless logic in diff calc

meowsbits committed:
consensus/ethash: add checks to tests, make diffcalc logic more intuitive

meowsbits committed:
core/rawdb: remove debugging line

meowsbits committed:
core,tests: implement use of parity chainspec for test chainconfigs

Failing:

$ go test -run TestBlockchain ./tests/...
--- FAIL: TestBlockchain (0.02s)
    --- FAIL: TestBlockchain/TransitionTests/bcEIP158ToByzantium/ByzantiumTransition.json (0.17s)
        block_test.go:49: Block #5 insertion into chain failed: invalid merkle root (remote: fa8bb175564fa6e8b8990d582fd009c465a5be4688d6f3dd7d540a76fff9169f local: 9806bb0ae31fc8c581a668ce3570c7ac2a0f07eaec5b3b6a554165be5ae66145 161enabled:true)
FAIL
FAIL    github.com/ethereum/go-ethereum/tests   13.008s

meowsbits committed:
tests: fix polluting iterated test config pointers

meowsbits committed:
vendor: add github.com/etclabscore/eth-x-chainspec

meowsbits committed:
params: remove Social and Ethersocial Block fields, use reward map

meowsbits committed:
vendor: update github.com/etclabscore/eth-x-chainspec

meowsbits committed:
params: remove redundant diff and reward dict keys

meowsbits committed:
tests: use relative path to chainspecs, like testdata path var

meowsbits committed:
cmd,consensus/ethash,params,tests: replace EIP649,EIP1234 blocks w/ maps

Replaces these blocks which enabled Reward and Difficulty variations
according to fork features, and replaces them with corresponding
equivalent difficulty and block reward maps.

meowsbits committed:
tests: don't panic on nil tests

This was causing eth/tracers tests to fail mysteriously.

meowsbits committed:
Merge branch 'master' into feat/parity-chainspec-tests

meowsbits committed:
params: ensure blockreward map parsing counts upwards

Since the map is not necessarily sorted.

meowsbits committed:
tests: add network information to reported error

meowsbits committed:
tests: use blockreward schedule for ByzantiumToConstantinopleAt5 spec

meowsbits committed:
vendor: update etclabscore/eth-x-chainspec dep

meowsbits committed:
cmd/utils,params: add ChainConfig.NetworkID, with flag parsing

meowsbits committed:
core: implement Genesis reader from filepath

meowsbits committed:
cmd/utils,cmd/geth: implement --chainspec.parity

meowsbits committed:
cmd/utils: set bootnodes via --chainspec.parity

This is ugly. I don't like it.
It reads and Marshals the file twice.

Ideally params.ChainConfig should contain field
Bootnodes and we can get rid of another 8 global variables.

But that requires reading and editing the set-configuration
logic around node and p2p setup and that's just more
complex than this is worth right now.

But for the record, I don't like it.
Not one bit.

meowsbits committed:
vendor: govendor update github.com/etclabscore/eth-x-chainspec

meowsbits committed:
core,tests: migrated eth-x-chainspec parityspec logic and tests

This includes tests output in tests/chainspecs_out, which may want
to be removed from vc or not generated at all.

meowsbits committed:
tests: use core/ package logic instead of dep for tests

meowsbits committed:
cmd/geth,cmd/utils,core: implement --chainspec.parity flag

meowsbits committed:
Merge branch 'master' into feat/parity-chainspec-tests-kensington

meowsbits committed:
core,vendor: add kensington chainspec, use math.HexOrDecimal256 for acc bal type

meowsbits committed:
Revert "core,vendor: add kensington chainspec, use math.HexOrDecimal256 for acc bal type"

This reverts commit c4bc9f91400431f2a5517c392f81e5eea7c9e172.

meowsbits committed:
cmd/geth,cmd/utils,core: finalize implement --chainspec.parity

sorpaas commented on core/xspec_parity_test.go:
```diff
@@ -0,0 +1,233 @@
+
+			} else {
+				// debugging the fixme above
+				// t.Log(spew.Sdump(wantG), spew.Sdump(mg))
```
Would be better if we don't have commented out code. For example, maybe change this to `t.Debugf` or something.

meowsbits replied:
removed with https://github.com/multi-geth/multi-geth/pull/103/commits/bfbf54318553532bc41adf2c699309da162f5cb4

meowsbits committed:
core: remove commented debugging code

meowsbits committed:
core: rename fns to use ParityConfig prefix

Since these are package-wide fns.

meowsbits committed:
core: fix linting errors

https://travis-ci.org/multi-geth/multi-geth/jobs/544697954

meowsbits committed:
params: fix config compat error for eip100/649

For chains configuring these features independently, eg. Kensington, this
throws an error. This test was written when Byz blockreward and difficulty feature
implementatations were tightly coupled, and should have been removed.

meowsbits committed:
params: removed inapplicable compat tests

meowsbits committed:
tests: Fix typo
